Asio-Server-and-Client/queue.cpp

```diff
@@ -0,0 +1,118 @@
+#include <cmath>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+
+template<typename T, uint64_t SIZE = 4096, uint64_t MAX_SPIN_ON_BUSY = 40000000>
+class ConcurrentQueue {
+private:
+    static constexpr unsigned Log2(unsigned n, unsigned p = 0) {
+        return (n <= 1) ? p : Log2(n / 2, p + 1);
+    }
+
+    static constexpr uint64_t closestExponentOf2(uint64_t x) {
+        return (1UL << ((uint64_t) (Log2(SIZE - 1)) + 1));
+    }
+
+    static constexpr uint64_t mRingModMask = closestExponentOf2(SIZE) - 1;
+    static constexpr uint64_t mSize = closestExponentOf2(SIZE);
+
+    static const T mEmpty;
+
+    T mMem[mSize];
+    std::mutex mLock;
+    uint64_t mReadPtr = 0;
+    uint64_t mWritePtr = 0;
+
+public:
+    const T& pop() {
+        if (!peek()) {
+            return mEmpty;
+        }
+
+        std::lock_guard<std::mutex> lock(mLock);
+
+        if (!peek()) {
+            return mEmpty;
+        }
+
+        T& ret = mMem[mReadPtr & mRingModMask];
+
+        mReadPtr++;
+        return ret;
+    }
+
+    bool peek() const {
+        return (mWritePtr != mReadPtr);
+    }
+
+    uint64_t getCount() const {
+        return mWritePtr > mReadPtr ? mWritePtr - mReadPtr : mReadPtr - mWritePtr;
+    }
```

Comment:
I would create a member variable mCount and increment it in push method, and decrement in pop method